src/lua/datetime.lua

```diff
@@ -0,0 +1,605 @@
+    local hms = false
+
+    local dt = local_dt(self)
+    local y0 = ffi.new('int[1]')
+    local M0 = ffi.new('int[1]')
+    local d0 = ffi.new('int[1]')
+    builtin.tnt_dt_to_ymd(dt, y0, M0, d0)
+    y0, M0, d0 = y0[0], M0[0], d0[0]
+
+    local y = obj.year
+    if y ~= nil then
+        check_range(y, 1, 9999, 'year')
+        ymd = true
+    end
+    local M = obj.month
+    if M ~= nil then
+        check_range(M, 1, 12, 'month')
+        ymd = true
+    end
+    local d = obj.day
+    if d ~= nil then
+        check_range(d, 1, 31, 'day', -1)
+        ymd = true
+    end
+
+    local lsecs = local_secs(self)
+    local h0 = math_floor(lsecs / (24 * 60)) % 24
+    local m0 = math_floor(lsecs / 60) % 60
+    local sec0 = lsecs % 60
+
+    local h = obj.hour
+    if h ~= nil then
+        check_range(h, 0, 23, 'hour')
+        hms = true
+    end
+    local m = obj.min
+    if m ~= nil then
+        check_range(m, 0, 59, 'min')
+        hms = true
+    end
+    local sec = obj.sec
+    if sec ~= nil then
+        check_range(sec, 0, 60, 'sec')
+        hms = true
+    end
+
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
+                       bool2int(msec ~= nil)
+    if count_usec > 0 then
+        if count_usec > 1 then
+            error('only one of nsec, usec or msecs may defined simultaneously', 2)
+        end
+        if usec ~= nil then
+            check_range(usec, 0, 1e6, 'usec')
+            self.nsec = usec * 1e3
+        elseif msec ~= nil then
+            check_range(msec, 0, 1e3, 'msec')
+            self.nsec = msec * 1e6
+        elseif nsec ~= nil then
+            check_range(nsec, 0, 1e9, 'nsec')
+            self.nsec = nsec
+        end
+    end
+
+    local ts = obj.timestamp
+    if ts ~= nil then
+        local sec_int, fraction
+        sec_int, fraction = math_modf(ts)
+        -- if there is one of nsec, usec, msec provided
+        -- then ignore fraction in timestamp
+        -- otherwise - use nsec, usec, or msec
+        if count_usec == 0 then
+            nsec = fraction * 1e9
+        else
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
+        end
+
+        self.epoch = sec_int
+        self.nsec = nsec
+
+        return self
+    end
+
+    local offset0 = self.tzoffset
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, -720, 720, 'tzoffset')
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+    end
+
+    -- .hour, .minute, .second
+    if hms then
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
+    end
+
+    if offset ~= nil then
+        self.tzoffset = offset
+    end
```

Comment:
Consider this diff:
```Diff
@@ -394,16 +394,14 @@ end
 local function datetime_update_dt(self, dt, new_offset)
     local epoch = local_secs(self)
     local secs_day = epoch % SECS_PER_DAY
-    epoch = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY +
-            secs_day
+    epoch = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY + secs_day
     self.epoch = utc_secs(epoch, new_offset)
 end
 
 local function datetime_ymd_update(self, y, M, d, new_offset)
     if d < 0 then
         d = builtin.tnt_dt_days_in_month(y, M)
-    end
-    if d > 28 then
+    elseif d > 28 then
         local day_in_month = builtin.tnt_dt_days_in_month(y, M)
         if d > day_in_month then
             error(('invalid number of days %d in month %d for %d'):
@@ -512,9 +510,13 @@ local function datetime_set(self, obj)
 
     local offset0 = self.tzoffset
     local offset = obj.tzoffset
+    local res_offset
     if offset ~= nil then
         offset = get_timezone(offset)
         check_range(offset, -720, 720, 'tzoffset')
+        res_offset = offset
+    else
+        res_offset = offset0
     end
 
     if obj.tz ~= nil then
@@ -523,18 +525,15 @@ local function datetime_set(self, obj)
 
     -- .year, .month, .day
     if ymd then
-        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, res_offset)
     end
 
     -- .hour, .minute, .second
     if hms then
-        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
-    end
-
-    if offset ~= nil then
-        self.tzoffset = offset
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, res_offset)
     end
 
+    self.tzoffset = res_offset
     return self
 end
```

Reply:
1. applied
2. applied
3. not applied directly, but simplified a bit:

```diff
@@ -409,16 +415,14 @@ end
 local function datetime_update_dt(self, dt, new_offset)
     local epoch = local_secs(self)
     local secs_day = epoch % SECS_PER_DAY
-    epoch = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY +
-            secs_day
+    epoch = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY + secs_day
     self.epoch = utc_secs(epoch, new_offset)
 end
 
 local function datetime_ymd_update(self, y, M, d, new_offset)
     if d < 0 then
         d = builtin.tnt_dt_days_in_month(y, M)
-    end
-    if d > 28 then
+    elseif d > 28 then
         local day_in_month = builtin.tnt_dt_days_in_month(y, M)
@@ -526,12 +530,12 @@ local function datetime_set(self, obj)
 
-    local offset0 = self.tzoffset
     local offset = obj.tzoffset
     if offset ~= nil then
-        offset = get_timezone(offset)
+        offset = get_timezone(offset, 'tzoffset')
         check_range(offset, -720, 840, 'tzoffset')
     end
+    offset = offset or self.tzoffset
 
@@ -539,17 +543,15 @@ local function datetime_set(self, obj)
 
     -- .year, .month, .day
     if ymd then
-        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset)
     end
 
     -- .hour, .minute, .second
     if hms then
-        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset)
     end
 
-    if offset ~= nil then
-        self.tzoffset = offset
-    end
+    self.tzoffset = offset
 
```